Reland "[InstructionSimplify] Enhance simplifySelectInst() (#163453)"

CongzheUalberta opened the pull request:
This reverts commit f1c1063.

PR #163453 was merged and reverted since it exposed a crash from codegenprepare which is fixed in #164628. 

Since this is PR for CI feedback and also the root cause of the crash is unrelated, I have added the skip-precommit-approval label.

CongzheUalberta committed:
[InstCombinePHI] Enhance PHI CSE to remove redundant phis

Enhanced PHI CSE to eliminate redundant PHIs, which could clean up
the IR and open up opportunities for other passes such as loop
vectorization.

CongzheUalberta committed:
address reviewer's comments

CongzheUalberta committed:
Address reviewer's comments and simplify the code.

CongzheUalberta committed:
Re-implemented in InstSimplify.

CongzheUalberta committed:
Address reviewer's comments made on Oct 21st.

CongzheUalberta committed:
Address reviewer's comments made on Oct 21st, 2nd part.

CongzheUalberta committed:
Address comments made on Oct 22nd.

CongzheUalberta committed:
Merge branch 'main' into enhanced-phi-cse

nikic commented on llvm/test/Transforms/InstCombine/select_with_identical_phi.ll:
```diff
@@ -0,0 +1,489 @@
+; %same.as.v1 is a select with two phis %v1 and %phi.to.remove as the true
+; and false values, while %v1 and %phi.to.remove are actually the same.
+; Fold the selection instruction %same.as.v1 to %v1.
+define void @select_with_identical_phi(ptr %m, ptr %n, i32 %count) {
```
It's okay to have one test for the motivating case, but for the remaining tests, please make the tests minimal. The matched pattern is already quite complicated without all the unrelated instructions.

CongzheUalberta replied:
Thanks for the comment. I've reduced all the remaining tests to minimal.

CongzheUalberta committed:
Reduced test cases to minimal.

nikic commented on llvm/test/Transforms/InstCombine/select_with_identical_phi.ll:
```diff
@@ -0,0 +1,280 @@
+define void @select_with_identical_phi_2(i1 %cmp1, i1 %cmp2, float %x) {
+; CHECK-LABEL: @select_with_identical_phi_2(
+; CHECK: select
+; CHECK-NOT: select
```
Please do keep the generated check lines.

nikic commented on llvm/test/Transforms/InstCombine/select_with_identical_phi.ll:
Can you please also add a negative test where the select conditions are not the same?

CongzheUalberta committed:
Address comments made on Oct 23rd.

CongzheUalberta commented:
> This reverts commit [f1c1063](https://github.com/llvm/llvm-project/commit/f1c1063acb12c7ce7724a88bf93351dd76b561c4).
> 
> PR #163453 was merged and reverted since it exposed a crash from codegenprepare which is fixed in #164628.
> 
> Since this is PR for CI feedback and also the root cause of the crash is unrelated, I have added the skip-precommit-approval label.

Since this PR is to re-land #163453 and the [crash](https://github.com/llvm/llvm-project/pull/163453#issuecomment-3429922732) exposed from #163453 is unrelated and should be tracked separately, I'll land this PR if there is no other comment. Internally we had some needs to land this PR at the earliest convenience.

Does it sound okay to you? @nikic

nikic commented:
@CongzheUalberta We need to wait for the crash fix to land first. It doesn't matter whether this patch introduces the issue or only exposes a pre-existing issue, the end result is still a crash.

dtcxzyw commented:
FYI https://github.com/llvm/llvm-project/pull/164628 has been landed.
